feat(workflows): enhance workflow authentication and request handling

tomer-shvadron opened the pull request:
- Implement combined authentication guard for improved security
- Add utility to extract workflow ID from URL for request enrichment
- Refactor request handling to include workflow ID in requests when present


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added plugins to support updating end user information and fetching company information within collection flows.
  - Introduced new utility functions for building collection flow URLs and handling workflow IDs.
  - Enhanced authentication with a new workflow-based guard for improved security and flexibility.
  - Expanded data models to track actor information in workflow runtime data.

- **Improvements**
  - Updated multiple API endpoints and URL patterns to include workflow IDs for better traceability.
  - Enhanced integration with external plugins and improved request handling by including credentials in cross-origin requests.
  - Refined query logic and context management in the collection flow application for more robust workflow handling.
  - Improved test infrastructure with additional context and actor services.

- **Bug Fixes**
  - Adjusted authentication decorators and guards to ensure correct application of default authentication logic.
  - Fixed various minor code style and readability issues.

- **Chores**
  - Removed deprecated token-based authentication logic and related environment variables.
  - Updated and reorganized type definitions and interfaces for clarity and maintainability.
  - Cleaned up redundant or outdated code and configuration files.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tomer-shvadron committed:
feat(workflows): enhance workflow authentication and request handling

- Implement combined authentication guard for improved security
- Add utility to extract workflow ID from URL for request enrichment
- Refactor request handling to include workflow ID in requests when present

tomer-shvadron committed:
Merge branch 'dev' into bal-3892

jonathan-shemer commented on services/workflows-service/src/common/guards/combined-auth.guard.ts:
```diff
@@ -0,0 +1,81 @@
+    const authenticatedEntity = req.user as AuthenticatedEntity & { projectIds?: TProjectIds };
+
+    if (
+      req.isAuthenticated() ||
```
if you have an expired session, this would fallback to `token` - we don't want that i think. can we think of another solution?

or can we assert that going through the edit flow would never include the `token` for the `endUser`?

scenario:
1. i'm an agent, i want to edit a specific workflow.
2. i've spent 3 hours editing it (and having lunch).
3. my session-cookie is now expired (ttl set to 2 hours).
4. i click submit, and the backend now thinks i'm the `endUser` rather than myself.

tomer-shvadron replied:
Done

tomer-shvadron committed:
chore(env): remove default example token and related dependencies

- Delete VITE_DEFAULT_EXAMPLE_TOKEN from example environment file
- Remove related token checks from entrypoint script and access token provider
- Adjust request functions to manage workflow ID more efficiently

tomer-shvadron committed:
feat(auth): enhance customer query and combined auth guard logic

- Add wfIdToken to enable customer query if either token is present
- Refactor combined auth guard to streamline workflow ID validation logic
- Ensure a seamless authentication experience for users with different roles

tomer-shvadron committed:
test(auth): add unit tests for Combined Auth Guard

- Implement unit tests for session and token authentication scenarios
- Include tests for error cases handling unauthorized requests

tomer-shvadron committed:
Merge branch 'dev' into bal-3892

tomer-shvadron committed:
Merge branch 'dev' into bal-3892

jonathan-shemer commented on apps/backoffice-v2/src/common/components/molecules/CaseVideoGuide/CaseVideoGuide.tsx:
```diff
@@ -35,7 +35,7 @@ export const CaseVideoGuide: FunctionComponent<CaseVideoGuideProps> = ({
                 }}
               >
-                <Skeleton className="absolute inset-0 size-full" />
+                <Skeleton className="size-full absolute inset-0" />
```
can we not commit these css changes?

tomer-shvadron replied:
That shit snuck in and when I try to fix it locally, it auto-fixes it back. It is annoying but I don't have a normal solution for that, I think we should try and treat that as tech-debt or something like that

jonathan-shemer commented on apps/kyb-app/src/common/providers/AccessTokenProvider/AccessTokenProvider.tsx:
```diff
@@ -10,34 +9,46 @@ interface IAccessTokenProviderProps {
+      const previousToken = searchParams.get('wf-id');
+
+      if (previousToken !== wfIdToken) {
+        setSearchParams({ 'wf-id': wfIdToken });
```
`wfIdToken` is a token? there's something i don't understand

tomer-shvadron replied:
Improved naming - all instances of `wfIdToken` were changed to `workflowId`

jonathan-shemer commented on apps/kyb-app/src/common/utils/request.ts:
```diff
@@ -1,9 +1,17 @@
 
-export const request = ky.create({
+// Helper function to get workflow ID from URL
+const getWorkflowIdFromUrl = () => {
+  const urlParams = new URLSearchParams(window.location.search);
+
+  return urlParams.get('wf-id');
```
why did you choose to have this as a query-param rather than the path?
i.e. `collection-flow/cm8r65m1y00b011aiwo8upuz9?token=a8b84fd0-3cae-46fa-88b8-e2f64e34573c`

tomer-shvadron replied:
I mainly followed existing structure with `token`, I didn't think it's necessary to use it as part of the path (because you anyway need to add it to the query-param upon creating the server request, otherwise we would also need to update ALL endpoints... WDYT?

jonathan-shemer commented on services/workflows-service/src/common/guards/combined-auth.guard.ts:
```diff
@@ -0,0 +1,84 @@
+
+import { AuthenticatedEntity, TProjectIds } from '@/types';
+import { WorkflowService } from '@/workflow/workflow.service';
+import { WorkflowTokenService } from '@/auth/workflow-token/workflow-token.service';
+
+@Injectable()
+export class CombinedAuthGuard implements CanActivate {
```
what do you think about
```suggestion
export class WorkflowAuthGuard implements CanActivate {
```
as this seems very coupled to authenticating access to workflows.

jonathan-shemer commented on services/workflows-service/src/common/guards/combined-auth.guard.ts:
```diff
@@ -0,0 +1,84 @@
+
+  async canActivate(context: ExecutionContext) {
+    const req = context.switchToHttp().getRequest<Request>();
+    // const workflowId = typeof req.query['wf-id'] === 'string' ? req.query['wf-id'] : null;
```
i removed this comment in my brach

jonathan-shemer commented on apps/kyb-app/src/common/utils/request.ts:
```diff
@@ -13,11 +21,15 @@ export const request = ky.create({
-        request.headers.set('Authorization', `Bearer ${getAccessToken()}`);
+        request.headers.set(
+          'Authorization',
+          `Bearer ${getAccessToken('wf-id') ?? getAccessToken()}`,
```
why do you use the `wf-id` as if it's a `Bearer` token? shouldn't this just be an ID?
Authorization should be handled solely in the session cookie and this header should not be set.

tomer-shvadron replied:
We have to do that to support other existing middlewares (`auth-key.middleware.ts` and `admin-key.middleware.ts`). Without it they fail.

jonathan-shemer replied:
let's fix those middlewares to handle these cases

tomer-shvadron committed:
refactor(auth): rename wfIdToken to workflowId in AccessTokenProvider

- Replace all references of wfIdToken with workflowId for clarity
- Update context and utility functions to reflect the new naming
- Ensure consistency across hooks that utilize access tokens

jonathan-shemer commented:
Let's make sure to address `collection-flow.controller.ts`, it still uses `TokenAuthGuard` exclusively (and any others).

chesterkmr commented on apps/kyb-app/src/common/utils/request.ts:
```diff
@@ -67,3 +81,33 @@ export const request = ky.create({
+// Export wrapped instance
+export const request = {
+  ...instance,
+  get: (url: string, options?: Options) => instance.get(url, addWorkflowId(options)),
```
Instead of wrapping each request type wont it be possible do add this logic to beforeRequest interceptor?

jonathan-shemer replied:
or to where we're already adding `token` (or in case we can't for some reason, [line 89](https://github.com/ballerine-io/ballerine/pull/3199/files#diff-7b24fd04c76c8a115b5d65db5056ebdb54622ff503f48215392b07ad242899a0R89) )?

tomer-shvadron committed:
feat(workflow): add workflow ID to URL handling and access control

- Introduce a custom hook to extract workflow ID from URL parameters
- Update routing to require workflow ID for relevant endpoints
- Simplify and standardize API calls with workflow ID for enhanced context management

tomer-shvadron committed:
feat(workflow): implement workflow ID fetching and routing

- Add fetchWorkflowId function to retrieve workflow IDs from the server
- Modify AccessTokenProvider to navigate based on workflow ID and language
- Update routing to support workflow-specific paths and parameters

tomer-shvadron committed:
fix(guards): update token authentication logic and rename decorator

- Modify token checks to allow null tokens to pass through
- Rename token-auth.decorator.ts to use-token-auth.decorator.ts
- Ensure unauthorized access is properly managed in workflow guard

tomer-shvadron committed:
refactor(workflow): update workflow ID handling to use query parameters

- Change retrieval of workflowId to use query parameters instead of URL params
- Refactor relevant navigation and request functions to accommodate new approach
- Remove deprecated session authentication guard and update workflow guard logic

tomer-shvadron committed:
feat(api): add update end user functionality

- Implement updateEndUser API endpoint to modify user details
- Add UpdateEndUserPluginData type for end-user data structure
- Create updateEndUser plugin for integrating with collection flow context

jonathan-shemer committed:
feat(workflow-runtime-data): add actor columns to indicate in audit log

tomer-shvadron committed:
feat(collection-flow): add company information fetching plugin

- Implement fetchCompanyInformation function with data validation
- Refine updateEndUser function to utilize Zod schema for validation
- Add new plugin for fetching company information to plugins repository

tomer-shvadron committed:
refactor(collection-flow): streamline fetchCompanyInformation function

- Replace createQueryParamsString with searchParams for cleaner code
- Introduce FetchCompanyInformationResultSchema for improved validation
- Enhance handling of plugin data and error logging in fetchCompanyInformationPlugin

tomer-shvadron committed:
feat(workflow): update collection flow URL to include workflow ID

- Modify the construction of the collection flow URL to include workflow ID
- Ensure compatibility in multiple service and adapter files
- Update references in external URL generation for improved traceability

jonathan-shemer commented on apps/backoffice-v2/src/pages/Entity/components/Case/components/CaseOptions/hooks/useCaseOptionsLogic/mutations/useCopyCollectionFlowLinkMutation/useCopyCollectionFlowLinkMutation.tsx:
```diff
@@ -9,7 +9,7 @@ export const useCopyCollectionFlowLinkMutation = ({ workflow }: { workflow: TWor
       }
 
-      const url = `${workflow.context.metadata.collectionFlowUrl}?token=${workflow.context.metadata.token}`;
+      const url = `${workflow.context.metadata.collectionFlowUrl}/?workflowId=${workflow.id}&token=${workflow.context.metadata.token}`;
```
common?

jonathan-shemer commented on apps/kyb-app/src/common/utils/create-query-params-string.ts:
```diff
@@ -0,0 +1,15 @@
+export const createQueryParamsString = (
```
isn't there something built in that can handle url serialization?

tomer-shvadron committed:
feat(collection-flow): implement buildCollectionFlowUrl utility

- Add a new utility function to build collection flow URLs with parameters
- Update multiple components to use the new utility for URL generation
- Remove obsolete utility function for creating query params string

tomer-shvadron committed:
Merge branch 'dev' into bal-3892

tomer-shvadron committed:
fix(collect-flow): handle missing collection flow URL

- Add error handling for missing collection flow URL
- Update URL construction logic to optionally include the token

tomer-shvadron committed:
chore(release): bump versions and update dependencies

- Bump version numbers for multiple packages
- Update dependencies for @ballerine/common and @ballerine/workflow-sdk packages
- Maintain consistency across various change logs and package files
